mautic/mautic
# Send queued webhooks in reversed chronological order

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Hardly
| Related user documentation PR URL | /
| Related developer documentation PR URL | https://github.com/mautic/developer-documentation/pull/68
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Zapier requires to send queued webhook entities in reversed chronological order:

https://zapier.com/developer/documentation/v2/deduplication/

To keep BC, this PR adds 2 new options to configure the order. One is in the Mautic configuration and that option is taken as a default one for all webhooks. The default value of that field is ASC, so it will work as before.

Another option is to configure the order for every webhook. That's what we need for Zapier.

#### Steps to test this PR:
1. **Before you checkout:** Make sure you have webhooks configured to be queued instead of sent immediately.
2. Create a new webhook with some actions you can easily trigger multiple times. Like contact update for example. You may use my [testing script](https://gist.github.com/escopecz/0d9adba753300b88326d) to catch the whole webhook request. Request Bin will strip the longer request body so you won't be able to confirm the order.
3. Trigger the actions at least 3 times.
4. Run 'app/console mautic:webhooks:process` command to trigger.
5. Notice the contact's `dateModified` is oldest at the top and latest at the bottom.
6. Apply this PR.
7. Repeat 3. and 4. Notice the order is still the same = BC.
8. Try to change order in the Mautic configuration to Reverse Chronological.
9. Fire some events again - the order is from latest to oldest.
10. Change the order in your webhook to Chronological.
11. Fire some events again - the order is from oldest to latest. The webhook's config overwrote the global config.
12. Run new API Library tests: https://github.com/mautic/api-library/pull/127
